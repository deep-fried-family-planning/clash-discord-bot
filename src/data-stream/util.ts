import type {ServerClan} from '#src/data/index.ts';
import type {CreateScheduleCommandInput} from '@aws-sdk/client-scheduler';
import type {DeleteCommandInput, PutCommandInput, TransactWriteCommandInput, UpdateCommandInput} from '@aws-sdk/lib-dynamodb';
import type {ClanWar} from 'clashofclans.js';
import * as Data from 'effect/Data';

export type Terminal = Data.TaggedEnum<{
  Delete       : DeleteCommandInput;
  Put          : PutCommandInput;
  Update       : UpdateCommandInput;
  TransactWrite: TransactWriteCommandInput;
  Schedule     : CreateScheduleCommandInput;
}>;

export const Terminal = Data.taggedEnum<Terminal>();
