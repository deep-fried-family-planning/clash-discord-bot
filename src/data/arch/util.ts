import type {Discord} from 'dfx';
import * as Data from 'effect/Data';

export type RegistryCaller = {
  guild? : Discord.Guild;
  user?  : Discord.User;
  member?: Discord.GuildMember;
};

export type RegistrySuccess = {
  title      : string;
  description: string;
};

export class RegistryDefect extends Data.TaggedError('RegistryDefect')<{
  cause?: any;
}> {}

export class RegistryFailure extends Data.TaggedError('RegistryError')<{
  cause?: any;
}> {}

export class RegistryUserError extends Data.TaggedError('RegistryUserError')<{
  message?: string;
  cause?  : any;
}> {}

export class RegistryAdminError extends Data.TaggedError('RegistryAdminError')<{
  message?: string;
  cause?  : any;
}> {}
