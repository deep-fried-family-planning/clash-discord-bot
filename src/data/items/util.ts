import * as S from 'effect/Schema';

const ApiEmbedType = S.Enums({
  RICH       : 'rich',
  IMAGE      : 'image',
  VIDEO      : 'video',
  GIFV       : 'gifv',
  ARTICLE    : 'article',
  LINK       : 'link',
  POLL_RESULT: 'poll_result',
} as const);

const ApiEmbedProvider = S.Struct({
  name: S.optional(S.String),
  url : S.optional(S.String),
});

const ApiEmbedImage = S.Struct({
  url      : S.optional(S.String),
  proxy_url: S.optional(S.String),
  height   : S.optional(S.Number),
  width    : S.optional(S.Number),
});

export const ApiEmbed = S.Struct({
  type     : S.optional(ApiEmbedType),
  provider : S.optional(ApiEmbedProvider),
  image    : S.optional(ApiEmbedImage),
  thumbnail: S.optional(S.Struct({
    url      : S.optional(S.String),
    proxy_url: S.optional(S.String),
    height   : S.optional(S.Number),
    width    : S.optional(S.Number),
  })),
  video: S.optional(S.Struct({
    url      : S.optional(S.String),
    proxy_url: S.optional(S.String),
    height   : S.optional(S.Number),
    width    : S.optional(S.Number),
  })),
  author: S.optional(S.Struct({
    name          : S.String,
    url           : S.optional(S.String),
    icon_url      : S.optional(S.String),
    proxy_icon_url: S.optional(S.String),
  })),
  color      : S.optional(S.Number),
  title      : S.optional(S.String),
  url        : S.optional(S.String),
  description: S.optional(S.String),
  footer     : S.optional(S.Struct({
    text          : S.optional(S.String),
    icon_url      : S.optional(S.String),
    proxy_icon_url: S.optional(S.String),
  })),
  fields: S.optional(S.Array(S.Struct({
    name  : S.String,
    value : S.String,
    inline: S.optional(S.Boolean),
  }))),
  timestamp: S.optional(S.String),
});
