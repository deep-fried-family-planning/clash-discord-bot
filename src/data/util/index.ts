export enum ClanVerification {
  admin     = 0,
  elder     = 1,
  coleader  = 2,
  leader    = 3,
  developer = 4,
}

export enum PlayerVerification {
  none      = 0,
  admin     = 1,
  token     = 2,
  developer = 3,
}
