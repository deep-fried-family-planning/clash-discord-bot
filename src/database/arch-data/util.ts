import {E, S} from '#src/internal/pure/effect.ts';
import {decode, encode} from '@msgpack/msgpack';
import {DateTime, type Duration, pipe} from 'effect';
import {deflate, inflate} from 'pako';

export const Created = S.transformOrFail(
  S.DateTimeUtc,
  S.UndefinedOr(S.DateTimeUtcFromSelf),
  {
    decode: (dt) => E.succeed(dt),
    encode: (dt) => dt ? E.succeed(dt) : DateTime.now,
  },
);

export const Updated = S.transformOrFail(
  S.DateTimeUtc,
  S.UndefinedOr(S.DateTimeUtcFromSelf),
  {
    decode: (dt) => E.succeed(dt),
    encode: () => DateTime.now,
  },
);

export const TimeToLive = (duration: Duration.Duration) =>
  S.transformOrFail(
    S.DateTimeUtcFromNumber,
    S.UndefinedOr(S.DateTimeUtcFromSelf),
    {
      decode: (dt) => E.succeed(dt),
      encode: (dt) => dt
        ? E.succeed(dt)
        : pipe(
          DateTime.now,
          E.map(
            DateTime.addDuration(duration),
          ),
        ),
    },
  );

export const Decoded = S.transform(
  S.Undefined,
  S.Literal(true),
  {
    decode: () => true as const,
    encode: () => undefined,
  },
);

export const Upgraded = pipe(
  S.Union(S.Undefined, S.Boolean),
  S.transform(
    S.Union(S.Undefined, S.Boolean),
    {
      decode: (enc) => enc,
      encode: () => undefined,
    },
  ),
  S.optional,
);

export const SelectMetadata = <A, I, R>(value: S.Schema<A, I, R>) =>
  S.Struct({
    value      : value,
    label      : S.String,
    description: S.optional(S.String),
    default    : S.optional(S.Boolean),
    emoji      : S.optional(S.Struct({
      name: S.String,
    })),
  });

export const CompressedBinary = <A, I, R>(schema: S.Schema<A, I, R>) =>
  S.transform(S.String, schema, {
    decode: (enc) => {
      const buff = Buffer.from(enc, 'base64');
      const pako = inflate(buff);
      return decode(pako) as any;
    },
    encode: (dec) => {
      const pack = encode(dec);
      const pako = deflate(pack);
      return Buffer.from(pako).toString('base64');
    },
  });
