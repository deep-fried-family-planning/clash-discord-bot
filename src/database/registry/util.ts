import {Data} from 'effect';

export class RegistrationError extends Data.TaggedError('deepfryer/RegistrationError')<{
  issue: string;
}> {}

export class RegistrationDefect extends Data.TaggedError('deepfryer/RegistrationDefect')<{
  issue: string;
}> {}
