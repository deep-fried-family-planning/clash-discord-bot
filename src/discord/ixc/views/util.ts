import type {IxState} from '#src/discord/ixc/store/derive-state.ts';
import type {snflk} from '#src/discord/types.ts';
import type {EmbedField} from 'dfx/types';
import type {str, und} from '#src/internal/pure/types-pure.ts';
import {pipe} from '#src/internal/pure/effect.ts';
import {emptyKV, reduceKV} from '#src/internal/pure/pure-kv.ts';
import {reduceL} from '#src/internal/pure/pure-list.ts';


export const isAdmin = (s: IxState) => s.user_roles.includes(s.server?.admin as snflk);


export const toReferenceFields = (fields: Record<str, EmbedField | und>) => pipe(
    fields,
    reduceKV([] as EmbedField[], (fs, f) => {
        if (f) {
            fs.push(f);
        }
        return fs;
    }),
);


export const fromReferenceFields = (fields?: EmbedField[]) => fields
    ? pipe(
        fields,
        reduceL(emptyKV<str, EmbedField | und>(), (fs, f) => {
            fs[f.name] = f;
            return fs;
        }),
    )
    : {};
