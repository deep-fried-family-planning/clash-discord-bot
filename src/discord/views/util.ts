import type {St} from '#src/discord/store/derive-state.ts';
import type {snow} from '#src/discord/types.ts';
import {pipe} from '#src/internal/pure/effect.ts';
import {emptyKV, reduceKV} from '#src/internal/pure/pure-kv.ts';
import {reduceL} from '#src/internal/pure/pure-list.ts';
import type {str, und} from '#src/internal/pure/types-pure.ts';
import type {EmbedField} from 'dfx/types';


export const isAdmin = (s: St) => s.user_roles.includes(s.server?.admin as snow);


export const toReferenceFields = (fields: Record<str, EmbedField | und>) => pipe(
  fields,
  reduceKV([] as EmbedField[], (fs, f) => {
    if (f) {
      fs.push(f);
    }
    return fs;
  }),
);


export const toReferenceFieldssssss = (fields: Record<str, str | und>) => pipe(
  fields,
  reduceKV([] as EmbedField[], (fs, f, n) => {
    if (f) {
      fs.push({
        name : n.trim(),
        value: f.trim(),
      });
    }
    return fs;
  }),
);


export const fromReferenceFields = (fields?: EmbedField[]) => fields
  ? pipe(
    fields,
    reduceL(emptyKV<str, EmbedField | und>(), (fs, f) => {
      fs[f.name] = f;
      return fs;
    }),
  )
  : {};
