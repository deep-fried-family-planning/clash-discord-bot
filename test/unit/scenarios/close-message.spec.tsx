import {CloseMessage} from '#test/unit/components/close-message.tsx';
import {makeTestRuntime} from '#test/unit/util.ts';
import {it} from '@effect/vitest';
import * as E from 'effect/Effect';

const runtime = makeTestRuntime([CloseMessage], false);

it.effect('when closing', E.fn(function* () {
  const root = yield* runtime.synthesize(CloseMessage, {}, {});

  expect(root).toMatchSnapshot();

  const res = yield* runtime.respond({
    id            : '1236074574509117491',
    token         : 'respond1',
    application_id: 'app',
    user_id       : 'user',
    guild_id      : 'guild',
    message       : root,
    type          : 3,
    data          : {
      custom_id     : 'CloseButton',
      component_type: 2,
    },
  });

  expect(res).toMatchSnapshot();
  expect(runtime.createUpdate).not.toBeCalled();
  expect(runtime.createSource).not.toBeCalled();
  expect(runtime.createModal).not.toBeCalled();
  expect(runtime.deferSource).not.toBeCalled();
  expect(runtime.deferEdit).not.toBeCalled();
  expect(runtime.discard).not.toBeCalled();
  expect(runtime.deferUpdate).toBeCalled();
  expect(runtime.dismount).toBeCalled();
}));
