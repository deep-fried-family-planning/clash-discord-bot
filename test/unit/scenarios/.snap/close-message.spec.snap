// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when closing 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "CloseButton",
          "disabled": undefined,
          "emoji": undefined,
          "label": undefined,
          "style": 2,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "content": undefined,
  "embeds": [
    {
      "author": undefined,
      "color": undefined,
      "description": "Hello World!",
      "fields": undefined,
      "footer": undefined,
      "image": {
        "url": "https://dffp.org/s/eJxrXpSZssY5J7841Te1uDgxPXVpQVF-QXHDsuKSxOTs4iYUuQk39ZC5VrlQ2sAqMbkkMz-v2MrQCqzAqbSkJD_PymDiAQDKAiej",
      },
      "title": undefined,
      "url": undefined,
    },
  ],
  "flags": undefined,
}
`;

exports[`when closing 2`] = `null`;
